addon/content/es-modules/contact-manager.js

```diff
@@ -0,0 +1,172 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import { freshColor } from "./utils.js";
+
+class ContactManager {
+  constructor() {
+    this._cache = new Map();
+    this._colorCache = new Map();
+    // We may ask for the same contact twice in rapid succession. In this
+    // case, we don't want to do queries multiple times. Instead we want to wait
+    // for the first query to finish. So, we keep track of all active queries.
+    this._activeFetch = new Map();
+  }
+  async getContactFromNameAndEmail({ name, email }) {
+    // [name] and [email] are from the message header
+    email = (email + "").toLowerCase();
+    // Might change in the future... who knows? ...
+    let key = email;
+    if (this._cache.has(key)) {
+      if (name) {
+        this._enrichWithName(key, name);
+      }
+      return this._cache.get(key);
+    }
+    if (this._colorCache.has(key)) {
+      // It is in the color cache, so we know that we don't have an address
+      // book entry for it, so just form a contact from what we have.
+      return {
+        emails: [email],
+        color: this._colorCache.get(email),
+        _email: email,
+        _name: name,
+        _card: null,
+        _useCardName: null,
+        avatar: "chrome://messenger/skin/addressbook/icons/contact-generic.svg",
+      };
+    }
+
+    if (this._activeFetch.has(key)) {
+      // If there's an active fetch going on for this contact,
+      // caching, etc. will be taken care of by the process that spawned the
+      // fetch. Therefore, we can safely return the result of the promise directly.
+      const contact = await this._activeFetch.get(key);
+      return contact;
```

Comment:
Hmm, I think we have to be careful here. A different call to this, with the same email might have a different name with it. We still want to only do the contact fetch once, but we might not be returning the same contact object.

This is why if the cache has it, we still enrichWithName every time.

Reply:
I'm still really confused about this multiple-name-with-same-email stuff. Can you give me a concrete example of how this happens? Is it the query to `browser.contacts` that does it? If so, how does it know? Or is this information stored elsewhere?

Reply:
The multiple name/same email is generally from the perspective of the email being displayed, rather than what's in the address book, for example:

If someone has an email address example@example.com, then X might be sending to them as `foo <example@example.com>` and Y sending as `bar <example@example.com`. The names here might be subtly different, or they might be very different - but just because we displayed one first, we don't necessarily want to display this name all the time. It might be a different name is an important piece of information for the user.

However, this could apply to the situation where there's a contact stored and but without any entry in the name fields. That's likely a smaller case, but I think we should handle it regardless so that we're always displaying accurate information.

Reply:
I've just had a thought, not quite sure if it will work or not. Here's the idea: Contact Manager will only ever return details of the contacts themselves plus colour information. So something like:

```js
{
        emails: ["me@example.com", "me1@example.com"],
        color: "#FF0000",
        contactName: "Me", // DisplayName or FirstName + LastName (selected via PreferDisplayName)
        avatar: "chrome://messenger/skin/addressbook/icons/contact-generic.svg",
      };
```

We might need useCardName in here as well. 
The contact name is only the name from the card, not what was in the email that triggered this.

The UI then exclusively handles working out what's actually displayed for a contact.

If you want, we could separate that out to a follow-up, though I was thinking that might be easier than fixing the activeFetch issue.

Reply:
This is close to what I was suggesting of using a selector to do the final assembly of the contact. I think it's a good idea!